### ISSUE TITLE
Add voronoi grid

### ISSUE THREAD

**Issue by virgile-baudrot** [EcoJulia/NeutralLandscapes.jl#30]
The following lines provide a simple way to make grid landscape using voronoi tesselation. It's working but need to be reframe to fit package requirement.

Computing the voronoi is done using [VoronoiCells.jl](https://github.com/JuliaGeometry/VoronoiCells.jl) and then the rasterization is by using [ArchGDAL.jl](https://github.com/yeesian/ArchGDAL.jl).

Digging into both dependencies, I should be able to make things clearer.

~~~ julia
using VoronoiCells
using GeometryBasics
using ArchGDAL
using Distributions
using Plots

const VC=VoronoiCells
const GB=GeometryBasics
const AG=ArchGDAL
const DB=Distributions
~~~

I created 3 functions. The 2 first create the set of polygones

~~~ julia
function voronoitess(min_x, max_x, min_y, max_y, nCell::Int)
    rect = VC.Rectangle(GB.Point2(min_x, min_y), GB.Point2(max_x, max_y))
    points = [GB.Point2(rand(DB.Uniform(min_x, max_x)), rand(DB.Uniform(min_y, max_y))) for _ in 1:nCell]
    return (tess=VC.voronoicells(points, rect),rect,points)
end

vortess = voronoitess(0, 1, 0, 1, 100)

function polygonset(tess)
    tc = tess.tess.Cells
    poly = [ArchGDAL.createpolygon([(tc[i][j][1], tc[i][j][2]) for j in vcat(1:length(tc[i]),1)])
    for i in 1:length(tc)]
    return (poly, tess.rect, tess.points)
end

polyset = polygonset(vortess)
plot(polyset[1])
~~~

![image](https://user-images.githubusercontent.com/6065464/108365024-06361300-71f7-11eb-8e52-5384a87a8c0a.png)

And the last is making a rasterization providing some features:

~~~ julia

# Create a grid over the map
function rasterize(polygonset, nrow, ncol,feature)
    w = rect.UpperRight[1] - rect.LowerLeft[1]
    h = rect.UpperRight[2] - rect.LowerLeft[2]
    gridcell_x=w/ncol
    gridcell_y=h/nrow
    gridX=(rect.LowerLeft[1]:gridcell_x:(rect.UpperRight[1]-gridcell_x)) .+ (gridcell_x/2)
    gridY=(rect.LowerLeft[2]:gridcell_y:(rect.UpperRight[2]-gridcell_y)) .+ (gridcell_y/2)

    gridX_seq=repeat(gridX,inner=nrow)
    gridY_seq=repeat(gridY,outer=ncol)

    gridCell = fill(0,nrow,ncol)
    for j in 1:length(gridCell)
        for i in 1:length(polygonset[1])
            if AG.contains(polygonset[1][i],
                AG.createpoint((gridX_seq[j], gridY_seq[j]))
            ) 
            gridCell[j] = feature[i]
            end        
        end
    end
    return gridCell
end

feature = rand(0:3,length(polyset[1]))
nrow, ncol = 200, 10

gridCell = rasterize(polyset, nrow, ncol, feature)
Plots.heatmap(gridCell)
~~~

![image](https://user-images.githubusercontent.com/6065464/108365473-8bb9c300-71f7-11eb-9189-c3a94691042b.png)

**Comment by mkborregaard** [comment]
So is this a suggestion to add the functionality or to put this in the docs?
I wouldn't be keen to add an ArchGDAL dep to this repo. VoroniCells might be OK (it adds Colors, GeometryBasics, VoronoDelaunay, GeometricalPredicates and RecipesBase).

**Comment by virgile-baudrot** [comment]
Well I don't know what you can do with that 😄 . Just that voronoi tesselation would be nice to have so I propose to work on it if it's something you think is relevant.
Basically, Voronoi is only made from GeometryBasics and VoronoDelaunay, so I just need to remove dependency with ArchGDAL, finding a better algorithm to rasterize.

**Comment by rafaqz** [comment]
@virgile-baudrot you see my dependency concerns are widespread in the Julia world ;)

But this would be great, but without ArchGDAL.

**Comment by mkborregaard** [comment]
Just for clarity - isn't rasterized voronoi just nearest-neighbor from a grid to the set of points? We have nearestneighbors.jl as a dep already.

**Comment by mkborregaard** [comment]
E.g. 
```julia
using NearestNeighbors
pts = rand(1.:1000, 2, 30)
tr = KDTree(pts)
mat = [knn(tr, [i, j], 1)[1][1] for i in 1:1000, j in 1:1000]
```
The heatmap:
<img width="488" alt="Screenshot 2021-02-19 at 00 55 02" src="https://user-images.githubusercontent.com/8429802/108437292-4e802000-724d-11eb-9fb5-f64c76c04c13.png">

**Comment by tpoisot** [comment]
I think it's a good improvement if we don't need ArchGDAL - I think @mkborregaard suggestion makes sense!

**Comment by tpoisot** [comment]
Wait, so is it NearestNeighborElement? I've added an option to set various values of k, but for k=1, this is a Voronoi tessellation?

**Comment by mkborregaard** [comment]
I was wondering if it wasn't just that yes. But couldn't tell from the code last night - I was heading home late on the train and was beer coding 😅

**Comment by mkborregaard** [comment]
But voronoi is just a vector format nearest neighbor yes

**Comment by virgile-baudrot** [comment]
Oh great. I didn't know that. So fantastic if it's already there!

**Comment by tpoisot** [comment]
Let's be explicit and add a `VoronoiGrid` type or something, as a thin wrapper on `NearestNeighborElement`? @virgile-baudrot do you want to do this? Happy to give feedback.

**Comment by virgile-baudrot** [comment]
ok, I'll try to implement it and PR

**Comment by tpoisot** [comment]
This was closed by #46
